completion_aggregator/models.py

```diff
@@ -1,27 +1,202 @@
+          to the value of possible.
+        * possible (float): The total sum of the possible completion values of all descendant
+          blocks that are visible to the user. This should be a positive integer.
+
+        Return Value
+        ------------
+        (BlockCompletion, bool)
+            A tuple comprising the created or updated
+            BlockCompletion object and a boolean value indicating whether the
+            object was newly created by this call.
+
+        Raises
+        ------
+        TypeError:
+            If the wrong type is passed for the parameters.
+
+        ValueError:
+            If the value of earned is greater than possible.
+
+        django.core.exceptions.ValidationError:
+            If earned / possible results in a number that is less than 0 or
+            greater than 1 or any float is less than zero.
+
+        django.db.DatabaseError:
+            If there was a problem getting, creating, or updating the
+            BlockCompletion record in the database.  This will also be a more
+            specific error, as described at
+            https://docs.djangoproject.com/en/1.11/ref/exceptions/#database-exceptions.
+            IntegrityError and OperationalError are relatively common
+            subclasses.
+
+        """
+        self.validate(user, course_key, block_key)
+        if earned > possible:
+            raise ValueError(_('Earned cannot be greater than the possible value.'))
+        if possible > 0.0:
+            percent = earned / possible
+        else:
+            percent = 1.0
+        obj, is_new = self.update_or_create(
+            user=user,
+            course_key=course_key,
+            aggregation_name=aggregation_name,
+            block_key=block_key,
+            defaults={
+                'percent': percent,
+                'possible': possible,
+                'earned': earned,
+            },
+        )
+        return obj, is_new
+
 
 @python_2_unicode_compatible
-class AggregateCompletion(TimeStampedModel):
+class Aggregator(TimeStampedModel):
     """
-    TODO: replace with a brief description of the model.
+    Aggregators are blocks that contain other blocks, are not themselves completable.
+
+    They are considered 100% complete when all descendant blocks are complete.
     """
 
-    # TODO: add field definitions
+    user = models.ForeignKey(User)
+    course_key = CourseKeyField(max_length=255)
+    aggregation_name = models.CharField(max_length=255)
+    block_key = UsageKeyField(max_length=255)
+    earned = models.FloatField(validators=[validate_positive_float])
+    possible = models.FloatField(validators=[validate_positive_float])
+    percent = models.FloatField(validators=[validate_percent])
+
+    objects = AggregatorManager()
+
+    class Meta(object):
+        """
+        Metadata describing the Aggregator model.
+        """
+
+        index_together = [
+            ('user', 'aggregation_name', 'course_key', 'block_key'),
```

Comment:
Yes.  Databases can only use prefixes of indexes, so if I want to search for user/aggregation name across all courses, I can't use an index of (user, course_key, aggregation name.

Reply:
Cool.  Just checking - is there an anticipated query pattern that's going to rely on the (course_key, block_key, user, aggregation_name) ordering of the unique constraint?

Reply:
Thanks for making me take a closer look at this.  Once all the fields are included, this is equivalent to the primary unique_together constraint.  Any of the prefixes are distinct though.  I've revised the duplicate index to track (user, aggregation_name, course_key).  This will allow, for instance, a dashboard query showing a user's course-level completion for all courses (using user, aggregation_name), or all of a user's aggregations of a particular kind in a course (Show me all the chapter aggregations in the current course), which will be useful for a quick single-level overview of the course.